flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java

```diff
@@ -0,0 +1,193 @@
+    this.properties = properties;
+    this.projectedFields = projectedFields;
+    this.isLimitPushDown = isLimitPushDown;
+    this.limit = limit;
+    this.filters = filters;
+    this.readableConfig = readableConfig;
+  }
+
+  @Override
+  public void applyProjection(int[][] projectFields) {
+    this.projectedFields = new int[projectFields.length];
+    for (int i = 0; i < projectFields.length; i++) {
+      Preconditions.checkArgument(projectFields[i].length == 1,
+          "Don't support nested projection in iceberg source now.");
+      this.projectedFields[i] = projectFields[i][0];
+    }
+  }
+
+  private DataStream<RowData> createDataStream(StreamExecutionEnvironment execEnv) {
+    return FlinkSource.forRowData()
+        .env(execEnv)
+        .tableLoader(loader)
+        .properties(properties)
+        .project(getProjectedSchema())
+        .limit(limit)
+        .filters(filters)
+        .flinkConf(readableConfig)
+        .build();
+  }
+
+  private TableSchema getProjectedSchema() {
+    if (projectedFields == null) {
+      return schema;
+    } else {
+      String[] fullNames = schema.getFieldNames();
+      DataType[] fullTypes = schema.getFieldDataTypes();
+      return TableSchema.builder().fields(
+          Arrays.stream(projectedFields).mapToObj(i -> fullNames[i]).toArray(String[]::new),
+          Arrays.stream(projectedFields).mapToObj(i -> fullTypes[i]).toArray(DataType[]::new)).build();
+    }
+  }
+
+  @Override
+  public void applyLimit(long newLimit) {
+    this.limit = newLimit;
+  }
+
+  @Override
+  public Result applyFilters(List<ResolvedExpression> flinkFilters) {
+    List<ResolvedExpression> acceptedFilters = Lists.newArrayList();
+    List<Expression> expressions = Lists.newArrayList();
+
+    for (ResolvedExpression resolvedExpression : flinkFilters) {
+      Optional<Expression> icebergExpression = FlinkFilters.convert(resolvedExpression);
+      if (icebergExpression.isPresent()) {
+        expressions.add(icebergExpression.get());
+        acceptedFilters.add(resolvedExpression);
+      }
+    }
+
+    this.filters = expressions;
+    return Result.of(acceptedFilters, flinkFilters);
+  }
+
+  @Override
+  public boolean supportsNestedProjection() {
+    // TODO: support nested projection
+    return false;
+  }
+
+  @Override
+  public ChangelogMode getChangelogMode() {
+    return ChangelogMode.insertOnly();
+  }
+
+  @Override
+  public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderContext) {
+    return new DataStreamScanProvider() {
+
+      // UPDATED - Needs to be added as support for other signature is entirely removed.
+      //           This should probably be ported to 1.14 as well to make future changes
+      //           easier to backport.
+      @Override
+      public DataStream<RowData> produceDataStream(
+          ProviderContext providerContext, StreamExecutionEnvironment execEnv) {
+        return createDataStream(execEnv);
+      }
```

Comment:
New code. The funtion signature below was deprecated and no longer is supported.